04 controllers.txt

```diff
@@ -0,0 +1,243 @@
+CONTROLLERS-------------------------------------------------------------------------------------------
+- Phoenix controllers act as intermediary modules
+- Controller functions (actions) are invoked from the router in reponse to HTTP requests
+- Actions perform necessary processes/data gathering before invoking view to render template OR returning JSON response
+
+- A newly generated Phoenix app has a single controller (PageController)
+
+ACTIONS----------------------------------------------------------------------------------------------
+- Controller actions are functions
+- Must match a route defined in the router
+
+Ex) > get "/", PageController, :index
+    - index is a function defined in PageController
+    - If we want to change this to...
+        i.e. > get "/", PageController, :test
+    - ... then we'd need to change the function name from index to test in PageController:
+
+    defmodule HelloWeb.PageController do
+        use HelloWeb, :controller
+
+        def index(conn, _params) do
+            render(conn, "index.html")
+        end
+    end
+
+
+- Conventions for action names:
+    index - renders a list of all items of the given resource type
+    show - renders an individual item by id
+    new - renders a form for creating a new item
+    create - receives params for one new item and saves it in a datastore
+    edit - retrieves an individual item by id and displays it in a form for editing
+    update - receives params for one edited item and saves it to a datastore
+    delete - receives an id for an item to be deleted and deletes it from a datastore
+Parameters
+- First is always conn (struct which holdes info about request i,e, host, path elements, etc)
+- Second is params (map that holds any parameters passed along in the HTTP request)
+    - Good practice to pattern match against params in the function signature to provide data in a simple package we can pass on to rendering
+    - If an action doesn't need parameter, you can use '_params' to name it
```

Comment:
Any argument can be preceded by an `_` so it doesn't get assigned. Alternatively, you can use `_` by itself, but it's better practice to have it precede a name so it provides context of what is being ignored.